### DISCUSSION TITLE
Ping not returning correct value

### DISCUSSION THREAD

**Discussion by georgeinva2004** [zwave-js/ZWaveJS.NET#44] [General]
Ping was always returning a "false" in the CMDResult.  Did some digging and noticed that the ZWaveNode.Ping() method had this to set the payload value.  
```
if (Res.Success)
{
   Res.SetPayload(JO.Value<bool>("responded"));
}
```

The actual response JSON looks like this:

```
{
  "type": "result",
  "success": true,
  "messageId": "c0f3d7b8-b9de-4794-9d83-323221c0939b",
  "result": {
    "responded": true
  }
}
```

I think the index needs to be "result.responded" not "responded" as currently implemented, so I changed it :).  BUT, that didn't work either.  I ended up rewriting it a bit to use the Newtonsoft JObject.SelectToken method which is providing the correct value for "responded".   The new code looks like this:

```
Driver.Instance.Callbacks.Add(ID, (JO) =>
{
    string jostr = JO.ToString();
    CMDResult Res = new CMDResult(JO);
    if (Res.Success)
    {
        bool responded = (bool)JO.SelectToken("result.responded");
        Res.SetPayload(responded);
        
    }
    
    Result.SetResult(Res);
});
```

I'm not sure if it's a bug in the Newtonsoft code or something else.  Based on the docs, the original code should have worked, but it's not for some reason.  I'm targeting .NET 7 in my application, if that makes a difference.

**Comment by marcus-j-davies** [comment]
Thanks @georgeinva2004, i'll fix that along with the key change today.
I must admit, despite using Newtonsoft.JSON for years, I still am confused between:

`JObject.value<T>(key)` and` JObject.SelectToken(path)`

If you know what the difference is, and why one is needed compared to the other, I'm dying to find out

A little tip on multiline code in Github - use 3 back ticks (I edited your comment)😄 
````
```
Code Here
```
````



**EDIT**
It seems `JObject.value<T>(key)` is used for a single key in the JSON Object, whereas `JObject.SelectToken(path)` is used to navigate to a value further into the Object - which would kind of make sense with your finding.

I'll scan though them today, and get it optimised

**Comment by georgeinva2004** [comment]
I'm in the same boat as you. I wish I could tell you. The docs say that it should work the way you have it implemented.  And thanks for the code-insertion tip.  I thought just clicking the button was enough :)

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
